server/aws-lsp-codewhisperer/src/language-server/chat/chatController.ts

```diff
@@ -67,108 +67,123 @@ export class ChatController implements ChatHandlers {
             return new ResponseError<ChatResult>(ErrorCodes.InternalError, sessionResult.error)
         }
 
-        const metric = new Metric<CombinedConversationEvent>({
-            cwsprChatConversationType: 'Chat',
-        })
+        return this.#withLspCancellation<ChatResult>(params.tabId, token, async checkIsCancelled => {
+            const metric = new Metric<CombinedConversationEvent>({
+                cwsprChatConversationType: 'Chat',
+            })
 
-        const triggerContext = await this.#getTriggerContext(params, metric)
-        const isNewConversation = !session.sessionId
+            const triggerContext = await this.#getTriggerContext(params, metric)
 
-        token.onCancellationRequested(() => {
-            this.#log('cancellation requested')
-            session.abortRequest()
-        })
+            // return empty result since the other promise will have been resolved
+            if (checkIsCancelled()) {
+                return {}
+            }
+
+            const isNewConversation = !session.sessionId
+
+            let response: GenerateAssistantResponseCommandOutput
+
+            const conversationIdentifier = session?.sessionId ?? 'New session'
+            try {
+                this.#log('Request for conversation id:', conversationIdentifier)
+                const requestInput = TriggerContextExtractor.getChatParamsFromTrigger(params, triggerContext)
+
+                metric.recordStart()
+                response = await session.generateAssistantResponse(requestInput)
+                this.#log('Response for conversationId:', conversationIdentifier, JSON.stringify(response.$metadata))
+            } catch (err) {
+                if (isObject(err) && 'name' in err && err.name === 'AbortError') {
+                    throw new CancellationError('Q api request aborted')
```

Comment:
here and down below, why do we throw cancellation error instead of returning an error object like we do in other error scenarios?

Reply:
I added an upstream error handling in `withLspCancellation`. I guess we should probably do one of the other